### ISSUE TITLE
stderr is printed as stdout on host

### ISSUE THREAD

**Issue by letmaik** [lsds/sgx-lkl#219]
Any output to stderr within the enclave is forwarded to the host and printed as stdout. Should this not stay stderr?

**Comment by prp** [comment]
Yes, there may be an issue how the virtio console device is set up and whether it correctly distinguishes between stdout/stderr.

**Comment by davidchisnall** [comment]
This is not directly connected to VirtIO. The first process created in Linux (or other *nix) system has all three of the first file descriptors opened to point to the default console device. The PTY layer has no idea whether a particular write comes from a file descriptors that the process thinks is stdout or stderr (it could find out by comparing against the fd number of the fd causing the write, but this would be a horrible layering violation). 

The distinction exists in UNIX so that commands can be chained together connecting stdout from one to a pipe that is set to stein for the next, with control and status messages (by convention, written to stderr) remaining on the console. The ability to redirect stderr was not found in early UNIX shells,
 at all, because the convention was that stderr was always for humans, whereas stdout may be next program to consume. 

If we want to support sending stderr somewhere else, then this is possible, with a bit of work. He VirtIO spec supports multiple channels for a console device. We could add a second channel and have user space close stderr and then open this device. The back end driver on the host could then write this to the host’s stderr. That is probably worth doing at some point, but it is not urgent unless we have a compelling use case.

**Comment by davidchisnall** [comment]
Removing the bug label, because this behaviour is entirely consistent with all other virtualisation platforms. Running a program in QEME, KVM, or whatever will see all console output sent to whatever fd 1 is for the process attached as the console driver.

**Comment by letmaik** [comment]
OK, makes sense. I note though that from the perspective of shift-and-lifting a Docker container it would change semantics and may be unexpected. Having said that, I agree that there's currently no compelling use case for this since logging output in release mode would not go to the host anyway. So as long as it's possible to redirect stdout/stderr separately within the enclave to somewhere (two files, for example) then it should be fine. Redirecting stderr to stdout should also be possible.

**Comment by davidchisnall** [comment]
Yup, redirecting output should be fine. I would like to have a separate console channel that we can use for stderr in the future though. It’s probably not too much effort, if you want to have a go...
